### PR TITLE
sessionctx/variable: set log bin

### PR DESCRIPTION
**Pull request by yangwenmai**
### What problem does this PR solve? 

impl https://github.com/pingcap/tidb/issues/9201
>3.Support close or enable binlog by set log_bin

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported variable/fields change

### CODE REVIEW

**Comment by morgo** [comment]
I have a request for MySQL compatibility:
- `log_bin` can work on a session/global basis
- `sql_log_bin` should work on a session-only basis (`sql_log_bin` should not be available globally at all.. due to history).
- Updates to one variable name should be reflected in the other variable (like an alias).

**Comment by yangwenmai** [comment]
now, my env go is 1.11.5, but ci is go 1.11.3.

**Comment by yangwenmai** [comment]
@WangXiangUSTC PTAL

File: sessionctx/binloginfo/binloginfo.go [hunk 1]
```diff
@@ -55,6 +58,40 @@ func GetPumpsClient() *pumpcli.PumpsClient {
 	return client
 }
 
+// GetOrCreatePumpsClient get or create the pumps client instance
+func GetOrCreatePumpsClient() *pumpcli.PumpsClient {
+	pumpsClientLock.Lock()
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: we can use read lock here to get the client, if client is nil, then lock and create client.

File: sessionctx/binloginfo/binloginfo.go [hunk 2]
```diff
@@ -55,6 +58,40 @@ func GetPumpsClient() *pumpcli.PumpsClient {
...
+	client := GetPumpsClient()
+	if client == nil {
+		var err error
+
+		securityOption := pd.SecurityOption{
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: some code is same with function `setupBinlogClient` in main.go, can we define a function to  do these things and then call it?

File: sessionctx/binloginfo/binloginfo.go (hunk 1 above)
Comment (yangwenmai) [reply sessionctx/binloginfo/binloginfo.go]: Done.

File: sessionctx/binloginfo/binloginfo.go (hunk 2 above)
Comment (yangwenmai) [reply sessionctx/binloginfo/binloginfo.go]: Done.

File: sessionctx/binloginfo/binloginfo.go [hunk 3]
```diff
@@ -57,6 +59,54 @@ func GetPumpsClient() *pumpcli.PumpsClient {
 	return client
 }
 
+// GetOrCreatePumpsClient get or create the pumps client instance
+func GetOrCreatePumpsClient() *pumpcli.PumpsClient {
+	pumpsClientLock.RLock()
+	client := GetPumpsClient()
+	if client == nil {
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: if two session call this function at the same time, and all get a nil client, will create pump client twice. 
how about add read lock when call GetPumpsClient, and if client is nil, use lock, and then create pump client.

File: session/session.go [hunk 4]
```diff
@@ -1398,7 +1398,14 @@ func createSession(store kv.Storage) (*session, error) {
 	domain.BindDomain(s, dom)
 	// session implements variable.GlobalVarAccessor. Bind it to ctx.
 	s.sessionVars.GlobalVarsAccessor = s
-	s.sessionVars.BinlogClient = binloginfo.GetPumpsClient()
+
+	tidbLogBin := variable.GetSysVar(variable.TiDBLogBin)
+	logBin := variable.GetSysVar(variable.LogBin)
+	if (logBin.Scope&variable.ScopeSession == 1 && logBin.Value == "1") ||
```
Comment (WangXiangUSTC) [inline session/session.go]: how abound define a function to judge whether write binlog?

File: sessionctx/binloginfo/binloginfo.go [hunk 5]
```diff
@@ -57,6 +59,54 @@ func GetPumpsClient() *pumpcli.PumpsClient {
...
+		var err error
+		client, err = CreatePumpsClient()
+		if err != nil {
+			return nil
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: need unlock

**Comment by yangwenmai** [comment]
@WangXiangUSTC PTAL

File: sessionctx/binloginfo/binloginfo.go [hunk 6]
```diff
@@ -49,14 +51,75 @@ type BinlogInfo struct {
...
 	pumpsClientLock.RUnlock()
 	return client
 }
 
+// GetPumpsClientByLogBin get the pumps client instance by log_bin variable.
+func GetPumpsClientByLogBin() *pumpcli.PumpsClient {
+	var binlogCli *pumpcli.PumpsClient
+	tidbLogBin := variable.GetSysVar(variable.TiDBLogBin)
+	logBin := variable.GetSysVar(variable.LogBin)
+	if (logBin.Scope&variable.ScopeSession == 1 && logBin.Value == "1") ||
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: there is a function to judge weather should write binlog, https://github.com/pingcap/tidb/blob/20463d6da90fdf12c0d9d18c15dc33a78334882d/sessionctx/binloginfo/binloginfo.go#L99

File: sessionctx/binloginfo/binloginfo.go [hunk 7]
```diff
@@ -49,14 +51,75 @@ type BinlogInfo struct {
...
+		atomic.StoreUint32(&variable.ProcessGeneralLog, 1)
+		binlogCli = GetOrCreatePumpsClient()
+	}
+	return binlogCli
+}
+
+// GetOrCreatePumpsClient get or create the pumps client instance
+func GetOrCreatePumpsClient() *pumpcli.PumpsClient {
+	client := GetPumpsClient()
+	if client == nil {
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: you can take a look 

https://github.com/pingcap/tidb-tools/blob/a17aa9c690a7a055a5e47d1c8295ca5f55f46ef3/tidb-binlog/pump_client/pump.go#L135

File: sessionctx/binloginfo/binloginfo.go [hunk 8]
```diff
@@ -49,14 +51,75 @@ type BinlogInfo struct {
...
+		var err error
+		pumpsClientLock.RLock()
+		client, err = CreatePumpsClient()
+		pumpsClientLock.RUnlock()
+		if err != nil {
+			return nil
+		}
+
+		err = pumpcli.InitLogger(config.GetGlobalConfig().Log.ToLogConfig())
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: how about move this line to `CreatePumpsClient` function

**Comment by yangwenmai** [comment]
@WangXiangUSTC PTAL

File: sessionctx/binloginfo/binloginfo.go [hunk 9]
```diff
@@ -49,14 +51,82 @@ type BinlogInfo struct {
...
+			client = pumpsClient
+		} else {
+			var err error
+			client, err = CreatePumpsClient()
+			if err != nil {
+				return nil, err
+			}
+			SetPumpsClient(client)
+		}
+		pumpsClientLock.RUnlock()
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: you means Unlock()?

File: sessionctx/binloginfo/binloginfo.go [hunk 10]
```diff
@@ -49,14 +51,82 @@ type BinlogInfo struct {
...
+	client := GetPumpsClient()
+	pumpsClientLock.RUnlock()
+	if client == nil {
+		pumpsClientLock.Lock()
+		if pumpsClient != nil {
+			client = pumpsClient
+		} else {
+			var err error
+			client, err = CreatePumpsClient()
+			if err != nil {
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: need unlock before return

File: sessionctx/binloginfo/binloginfo.go [hunk 11]
```diff
@@ -49,14 +51,82 @@ type BinlogInfo struct {
...
 	pumpsClientLock.RUnlock()
 	return client
 }
 
+// GetPumpsClientByLogBin get the pumps client instance by log_bin variable.
+func GetPumpsClientByLogBin() (*pumpcli.PumpsClient, error) {
+	var binlogCli *pumpcli.PumpsClient
+	var err error
+	if ShouldEnableBinlog() {
+		atomic.StoreUint32(&variable.ProcessGeneralLog, 1)
```
Comment (WangXiangUSTC) [inline sessionctx/binloginfo/binloginfo.go]: why change this variable?

**Comment by WangXiangUSTC** [comment]
@yangwenmai don't need support set log_bin now, I will close this pr. Thanks for your contribution

**Pull request closed without merging**